src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs

```diff
@@ -0,0 +1,554 @@
+                });
+            }
+
+            if (request.HasScope(OpenIdConnectConstants.Scopes.OfflineAccess) && !application.AllowRefreshTokenFlow)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.InvalidClient,
+                    ErrorDescription = T["Offline scope is not allowed for this OpenID Connect Application"]
+                });
+            }
+
+            if (request.IsAuthorizationCodeFlow() && !application.AllowAuthorizationCodeFlow)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                    ErrorDescription = T["Authorization Code Flow is not allowed for this OpenID Connect Application"]
+                });
+            }
+
+            if (request.IsImplicitFlow() && !application.AllowImplicitFlow)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                    ErrorDescription = T["Implicit Flow is not allowed for this OpenID Connect Application"]
+                });
+            }
+
+            if (request.IsHybridFlow() && !application.AllowHybridFlow)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
+                    ErrorDescription = T["Hybrid Flow is not allowed for this OpenID Connect Application"]
+                });
+            }
+
+            if (application.SkipConsent)
+            {
+                return await IssueAccessIdentityTokens(request);
+            }
+
+            return View(new AuthorizeViewModel
+            {
+                ApplicationName = application.DisplayName,
+                RequestId = request.RequestId,
+                Scope = request.Scope
+            });
+        }
+
+        [ActionName(nameof(Authorize))]
+        [HttpPost, FormValueRequired("submit.Accept")]
+        public async Task<IActionResult> Accept()
+        {
+            var response = HttpContext.GetOpenIdConnectResponse();
+            if (response != null)
+            {
+                return View("Error", new ErrorViewModel
+                {
+                    Error = response.Error,
+                    ErrorDescription = response.ErrorDescription
+                });
+            }
+
+            // If the request is missing, this likely means that
+            // this endpoint was not enabled in the settings.
+            // In this case, simply return a 404 response.
+            var request = HttpContext.GetOpenIdConnectRequest();
+            if (request == null)
+            {
+                return NotFound();
+            }
+
+            return await IssueAccessIdentityTokens(request);
```

Comment:
In theory, I'd say yes, but in this case, you'd have to wrap all the other returned values into a `Task.FromResult(...)` call, which would make the code ugly (for a marginal perf' gain).

E.g:

```csharp
return Task.FromResult<IActionResult>(View("Error", new ErrorViewModel
{
    Error = OpenIdConnectConstants.Errors.UnauthorizedClient,
    ErrorDescription = T["Implicit Flow is not allowed for this OpenID Connect Application"]
}));
```